v4/core/authenticator.go

```diff
@@ -21,6 +21,16 @@ import (
 // Authenticator describes the set of methods implemented by each authenticator.
 type Authenticator interface {
 	AuthenticationType() string
-	Authenticate(*http.Request) error
+	Authenticate(*http.Request) *AuthenticationError
```

Comment:
I think we can revert this back to `error` as the return type in order to maintain compatibility in the API.   This will mean that where we call Authenticate() in BaseService.Request(), you'll need to adjust that code to cast the "error" to an "AuthenticationError" when trying to retrieve the DetailedResponse field.   I hadn't thought of this "trick" before, but should have :)